### ISSUE TITLE
Support access the DevServer instance from a Webpack plugin

### ISSUE THREAD

**Issue by fightwithtiger** [webpack/webpack-dev-server#5025]
## Feature request

**What is the expected behavior?**
Is it possible to access DevServer instances from the Webpack plug-in somehow? I want to send custom events from server to client.

This is totally [possible with Vite.js](https://vitejs.dev/guide/api-plugin.html#client-server-communication), but wonder if there is also a way with Webpack.

**What is motivation or use case for adding/changing the behavior?**

**How should this be implemented in your opinion?**
A new hook is provided in the plugin to access devServer

**Are you willing to work on this yourself?**

**Comment by Basavarajrp** [comment]
Hey @fightwithtiger 👋,
I was just going through the issues section and found this interesting, also checked the Vite approach in the link that u shared. 
I just wanted to know the use case of this feature. How this will be helpful for the developer experience so that I can explore more and try to think of a solution.

Happy Coding 🧑‍💻
-- Basavaraj Patil.

**Comment by Anubhab2003** [comment]
Hey I am Interested in contributing this issue, How to get started ?

**Comment by fightwithtiger** [comment]
> Hey @fightwithtiger 👋, I was just going through the issues section and found this interesting, also checked the Vite approach in the link that u shared. I just wanted to know the use case of this feature. How this will be helpful for the developer experience so that I can explore more and try to think of a solution.
> 
> Happy Coding 🧑‍💻 -- Basavaraj Patil.

Thank you for your reply. I haven't checked my messages for a long time. 
For example: [link](https://github.com/fightwithtiger/sw-heart-beat/blob/main/src/plugin.ts)
As indicated by the function of this Vite plugin, I want to inject a static JS resource after the project is built. If it is in the dev stage, the Vite configureServer API can be used to obtain the server, and then hijack the request to rewrite it as the specified static JS content. However, it seems that there is no good way for me to achieve this function in the dev stage of webpack.

**Comment by fightwithtiger** [comment]
> Hey I am Interested in contributing this issue, How to get started ?

Thank you for your reply. I haven't checked my messages for a long time. 
For example: [link](https://github.com/fightwithtiger/sw-heart-beat/blob/main/src/plugin.ts)
As indicated by the function of this Vite plugin, I want to inject a static JS resource after the project is built. If it is in the dev stage, the Vite configureServer API can be used to obtain the server, and then hijack the request to rewrite it as the specified static JS content. However, it seems that there is no good way for me to achieve this function in the dev stage of webpack.

**Comment by alexander-akait** [comment]
You can use middlewares to change something in request

**Comment by fightwithtiger** [comment]
> You can use middlewares to change something in request
Is it written like this?

```
// webpack.config.js
const path = require('path');

module.exports = {
  mode: 'development',
  entry: './src/index.js',
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: 'bundle.js',
    publicPath: '/',
  },
  devServer: {
    static: {
      directory: path.join(__dirname, 'dist'),
    },
    port: 3000,
    setupMiddlewares: (middlewares, devServer) => {
      if (!devServer) {
        throw new Error('webpack-dev-server is not defined');
      }

      middlewares.unshift({
        name: 'custom-middleware',
        path: '/',
        middleware: (req, res, next) => {
          if (req.url.endsWith('.js')) {
            const modifiedContent = 'console.log("This is modified by custom middleware!");';
            res.setHeader('Content-Type', 'application/javascript');
            res.send(modifiedContent);
          } else {
            next();
          }
        },
      });

      return middlewares;
    },
  },
};
```

**Comment by alexander-akait** [comment]
Yes, you don't need `!devServer`, `devServer` is always will be not undefined
